### PR TITLE
feat: option for scaling down resolution on move

### PR DESCRIPTION
**Pull request by haakonflatval-cognite**
#### Type of change

Feat  

#### Jira ticket :blue_book:

https://cognitedata.atlassian.net/browse/REV-619

## Description :pencil:

We want to scale down resolution when camera is moving for low-end devices. This will allow better user experience, maybe.

This PR adds an option for scaling down the camera resolution when on the move. It is provided as a factor of the pixels to render when moving.

In addition, there is now a `setRendersizeThreshold` function exposed in viewer. This isn't really necessary for the resize-on-move feature, but was used in an earlier iteration of this PR, and might be useful for some. Can remove on demand.

## How has this been tested? :mag:

Manually

## Test instructions :information_source:

Open examples, move around

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [x] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the public documentation.
- [x] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [x] I have refactored the code for readability to the best of my ability.
- [x] I have checked that my changes do not introduce regressions in the public documentation.
- [x] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [x] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [x] I have added TSDoc to any public facing changes.
- [x] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [x] I have noted down and am currently tracking any technical debt introduced in this PR.
- [x] I have thoroughly thought about the architecture of this implementation.
- [x] I have asked peers to test this feature.

### CODE REVIEW

**Comment by haakonflatval-cognite** [comment]
TODO: Adaptive point sizing seems to not be working.. Gonna check it out

File: viewer/packages/api/src/public/RevealOptions.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ export type RevealOptions = {
   continuousModelStreaming?: boolean;
   outputRenderTarget?: { target: THREE.WebGLRenderTarget; autoSize?: boolean };
   rendererResolutionThreshold?: number;
+  movingResolutionFactor?: number;
```
Comment (larsmoa) [inline viewer/packages/api/src/public/RevealOptions.ts:21]: Suggest this is renamed to `cameraResolutionFactor`. Also, add TSDoc - I also see that most other stuff here is missing and that we document a non-existant field.

File: viewer/packages/api/src/public/migration/Cognite3DViewer.ts [hunk 2]
```diff
@@ -333,6 +333,14 @@ export class Cognite3DViewer {
     });
   }
 
+  /**
+   * Set resolution threshold.
+   * @param threshold
+   */
+  setResolutionThreshold(threshold: number): void {
```
Comment (larsmoa) [inline viewer/packages/api/src/public/migration/Cognite3DViewer.ts:340]: This means nothing to me. Is this what should be named `setMaximumRenderResolution`? Also, document how it works (i.e. that its width x height) - and what happens when resolution cap is exceeded.

File: viewer/packages/api/src/public/RevealOptions.ts (hunk 1 above)
Comment (larsmoa) [reply viewer/packages/api/src/public/RevealOptions.ts:21]: Or maybe this is purely internal? Should still be documented, but not that important

File: viewer/packages/api/src/public/migration/types.ts [hunk 3]
```diff
@@ -79,6 +79,13 @@ export interface Cognite3DViewerOptions {
    */
   rendererResolutionThreshold?: number;
 
+  /**
+   * A factor that will scale down the resolution when moving the camera. This can
+   * be used to achieve a better user experience on devices with limited hardware.
+   * A value of e.g. 0.25 will approximately divide the number of pixels on the screen by four.
+   */
+  movingResolutionFactor?: number;
```
Comment (larsmoa) [inline viewer/packages/api/src/public/migration/types.ts:87]: As above, I suggest to make this `cameraMovingResolutionFactor` or something. Enforce values between 0 and 1.

File: viewer/packages/rendering/src/ResizeHandler.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+import { WebGLRenderer, PerspectiveCamera, Vector2 } from 'three';
+
+import { CameraManager } from '@reveal/camera-manager';
+
+export type ResizeHandlerOptions = {
+  renderResolutionThreshold?: number;
+  movingResolutionFactor?: number;
+};
+
+export class ResizeHandler {
```
Comment (larsmoa) [inline viewer/packages/rendering/src/ResizeHandler.ts:14]: Nice! Would be even nicer with a test ;)

File: viewer/packages/rendering/src/pointcloud-rendering/PointCloudMaterial.ts [hunk 5]
```diff
@@ -335,37 +334,25 @@ export class PointCloudMaterial extends RawShaderMaterial {
...
+    this.screenWidth = renderSize.x;
+    this.screenHeight = renderSize.y;
   }
 
   updateMaterial(
     octreeParams: OctreeMaterialParams,
     visibilityTextureData: Uint8Array,
     camera: Camera,
-    renderer: WebGLRenderer
+    _renderer: WebGLRenderer
```
Comment (larsmoa) [inline viewer/packages/rendering/src/pointcloud-rendering/PointCloudMaterial.ts:349]: Remove?

**Review by larsmoa — CHANGES_REQUESTED** [review]
See comments

File: viewer/packages/api/src/public/RevealOptions.ts (hunk 1 above)
Comment (haakonflatval-cognite) [reply viewer/packages/api/src/public/RevealOptions.ts:21]: You sure about the `cameraResolutionFactor` name? I believe it should somehow notify that we're talking about the resolution when the camera is moving.

File: viewer/packages/api/src/public/migration/Cognite3DViewer.ts [hunk 6]
```diff
@@ -333,6 +333,15 @@ export class Cognite3DViewer {
...
   }
 
+  /**
+   * Limit the number of pixels on the screen. When the render area has more pixels than this max resolution,
+   * Reveal will automatically render to an image with at most the max resolution, scaled up to fit the render area.
+   * @param maxResolution
+   */
+  setMaximumRenderResolution(maxResolution: number): void {
+    this._revealManagerHelper.revealManager.setResolutionThreshold(maxResolution);
+  }
```
Comment (larsmoa) [inline viewer/packages/api/src/public/migration/Cognite3DViewer.ts:343]: Should the constructor-option now be marked `@deprecated` for removal in 5.0, and refer to this setter instead?

File: viewer/reveal.api.md [hunk 7]
```diff
@@ -397,6 +398,7 @@ export interface Cognite3DViewerOptions {
         opacity: number;
     };
     logMetrics?: boolean;
+    movingResolutionFactor?: number;
```
Comment (larsmoa) [inline viewer/reveal.api.md:401]: Update docs

**Review by larsmoa — COMMENTED** [review]
See comments - also, EDL works shitty when reducing resolution, which we should fix...

File: viewer/packages/api/src/public/migration/Cognite3DViewer.ts (hunk 6 above)
Comment (haakonflatval-cognite) [reply viewer/packages/api/src/public/migration/Cognite3DViewer.ts:343]: Sure, I'll change it.

Now that I think of it.. The moving-resolution can also be set dynamically? Should we have a similar setter for that one? :thinking:

File: documentation/docs/examples/rendering-resolution.mdx [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+# Resolution Control
+
+Reveal will automatically limit its rendering resolution when renderering to a large surface for the sake of performance. This may result in unnecessarily low-detail renders on high-end devices with large monitors.
+
+The user can use the `ResolutionOptions` type to control the rendering resolution. `ResolutionOptions` has the following fields:
+
+| Field                          | Field type                  | Description                                                                      |
+| ------------------------------ | --------------------------- | ----------------------------------- |
+| `maxRenderResolution`          | `number?`                   | Set the maximum render resolution                                                |
+| `movingCameraResolutionFactor` | `number?`                   | Set the factor that the max resolution is multiplied with when camera is moving. |
```
Comment (larsmoa) [inline documentation/docs/examples/rendering-resolution.mdx:16]:
```suggestion
| `movingCameraResolutionFactor` | `number?`                   | Set the factor that the max resolution is multiplied with when camera is moving. Must be between 0 and 1.  Lower factor means lower resolution while camera is moving. |
```

File: viewer/packages/rendering/src/ResizeHandler.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+    assert(factor > 0 && factor <= 1);
+
+    if (factor !== 1) {
+      this._cameraManager.on('cameraChange', this._onCameraChangeCallback);
+      this._cameraManager.on('cameraStop', this._onCameraStopCallback);
+    } else {
+      this._cameraManager.off('cameraChange', this._onCameraChangeCallback);
+      this._cameraManager.off('cameraStop', this._onCameraStopCallback);
+    }
```
Comment (larsmoa) [inline viewer/packages/rendering/src/ResizeHandler.ts:80]: I see a bug. If you call this function twice with different (but not 1) factor, you will add more and more event listeners.

File: viewer/packages/rendering/src/ResizeHandler.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  public setMovingCameraResolutionFactor(factor: number): void {
+    this._shouldResize = true;
+
+    if (factor === this._movingResolutionFactor) {
+      return;
+    }
+
+    assert(factor > 0 && factor <= 1);
```
Comment (larsmoa) [inline viewer/packages/rendering/src/ResizeHandler.ts:72]: Replace with a proper `throw Error(<msg>)` - assert is only for stuff you think cannot happen and not to user errors. See https://stackoverflow.com/a/13832636. 

Also, add unit test for this.

File: viewer/packages/rendering/src/glsl/pointcloud/pointcloud.vert [hunk 11]
```diff
@@ -2,6 +2,7 @@ precision highp float;
 precision highp int;
 
 #define max_clip_boxes 30
+#define point_size_unit_in_screen_height (1.0 / 720.0)
```
Comment (larsmoa) [inline viewer/packages/rendering/src/glsl/pointcloud/pointcloud.vert:5]: Add a line about where this magic constant comes from

File: viewer/packages/rendering/src/glsl/pointcloud/edl.glsl [hunk 12]
```diff
@@ -4,10 +4,13 @@
 
 #define MAX_FLOAT 3.402823466e+38
 
+#define point_size_unit_per_screen_height 720.0
```
Comment (larsmoa) [inline viewer/packages/rendering/src/glsl/pointcloud/edl.glsl:7]: Add a line about where this magic constant comes from

**Review by larsmoa — CHANGES_REQUESTED** [review]
See comments

**Review by larsmoa — APPROVED** [review]

**Pull request merged**